File: medify-backend/api/src/main/java/medify/backend/api/dto/IntakeHistoryEntry.java
package medify.backend.api.dto;

import medify.backend.domain.model.Intake;
import medify.backend.domain.model.IntakeStatus;
import medify.backend.domain.model.Timing;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * One row of the patient/caregiver-facing intake history — a computed projection over an
 * existing {@link Intake}, not a persisted concept of its own. The mapping in {@link #from} is
 * the single place "what actually happened to this dose" is decided, so the patient app and the
 * caregiver view (both hitting the same /api/intakes/history endpoint) always agree.
 */
public record IntakeHistoryEntry(
        Long intakeId,
        Timing timing,
        LocalDate scheduledDate,
        LocalDateTime scheduledTime,
        IntakeStatus status,
        Outcome outcome,
        /** When the IR sensor actually confirmed pickup — null if never taken. */
        LocalDateTime takenAt,
        /** takenAt - scheduledTime, in minutes. Negative means taken early (during the early
         *  window, before the scheduled instant). Null if never taken. */
        Long latenessMinutes,
        /** Whether this occurrence was postponed at least once, regardless of the final outcome —
         *  e.g. still worth surfacing even when the outcome ended up TAKEN_AFTER_MISSED. */
        boolean wasPostponed
) {
    public enum Outcome {
        TAKEN_ON_TIME,
        TAKEN_AFTER_POSTPONED,
        TAKEN_AFTER_MISSED,
        MISSED,
        SKIPPED,
        INCOMPLETE,
        PENDING,
        APPROVED,
        POSTPONED,
        DISPENSING,
        DISPENSED
    }

    public static IntakeHistoryEntry from(Intake intake) {
        boolean wasPostponed = intake.getPostponedAt() != null;
        boolean wasMissed = intake.getMissedAt() != null;

        Outcome outcome = switch (intake.getStatus()) {
            case TAKEN -> wasMissed ? Outcome.TAKEN_AFTER_MISSED
                    : wasPostponed ? Outcome.TAKEN_AFTER_POSTPONED
                    : Outcome.TAKEN_ON_TIME;
            case MISSED -> Outcome.MISSED;
            case SKIPPED -> Outcome.SKIPPED;
            case INCOMPLETE -> Outcome.INCOMPLETE;
            case PENDING -> Outcome.PENDING;
            case APPROVED -> Outcome.APPROVED;
            case POSTPONED -> Outcome.POSTPONED;
            case DISPENSING -> Outcome.DISPENSING;
            case DISPENSED -> Outcome.DISPENSED;
        };

        LocalDateTime takenAt = intake.getReleasedTime();
        Long latenessMinutes = takenAt != null
                ? Duration.between(intake.getScheduledTime(), takenAt).toMinutes()
                : null;

        return new IntakeHistoryEntry(
                intake.getId(),
                intake.getTiming(),
                intake.getScheduledDate(),
                intake.getScheduledTime(),
                intake.getStatus(),
                outcome,
                takenAt,
                latenessMinutes,
                wasPostponed
        );
    }
}

File: medify_app/lib/models/intake_history_entry.dart
import 'medicine.dart';

/// Mirrors the backend's IntakeHistoryEntry.Outcome — the single computed classification of
/// "what happened to this dose," shared by both the patient app and the caregiver view so they
/// never disagree.
enum IntakeOutcome {
  takenOnTime,
  takenAfterPostponed,
  takenAfterMissed,
  missed,
  skipped,
  incomplete,
  pending,
  approved,
  postponed,
  dispensing,
  dispensed;

  static IntakeOutcome fromApi(String? value) {
    switch (value) {
      case 'TAKEN_ON_TIME':        return IntakeOutcome.takenOnTime;
      case 'TAKEN_AFTER_POSTPONED': return IntakeOutcome.takenAfterPostponed;
      case 'TAKEN_AFTER_MISSED':   return IntakeOutcome.takenAfterMissed;
      case 'MISSED':               return IntakeOutcome.missed;
      case 'SKIPPED':              return IntakeOutcome.skipped;
      case 'INCOMPLETE':           return IntakeOutcome.incomplete;
      case 'PENDING':               return IntakeOutcome.pending;
      case 'APPROVED':             return IntakeOutcome.approved;
      case 'POSTPONED':            return IntakeOutcome.postponed;
      case 'DISPENSING':           return IntakeOutcome.dispensing;
      case 'DISPENSED':            return IntakeOutcome.dispensed;
      default:                     return IntakeOutcome.pending;
    }
  }
}

/// One row of intake history, as returned by GET /api/intakes/history. `scheduledDate` isn't
/// parsed separately — the backend serializes a LocalDate as a 3-element array, which
/// [parseBackendDateTime] doesn't handle (it expects a 5+ element LocalDateTime array), so the
/// day for grouping is derived from [scheduledTime] instead.
class IntakeHistoryEntry {
  final int intakeId;
  final String timing; // MORNING / NOON / EVENING
  final DateTime scheduledTime;
  final String status; // raw backend IntakeStatus, for anything not covered by outcome
  final IntakeOutcome outcome;
  final DateTime? takenAt;
  final int? latenessMinutes;
  final bool wasPostponed;

  IntakeHistoryEntry({
    required this.intakeId,
    required this.timing,
    required this.scheduledTime,
    required this.status,
    required this.outcome,
    this.takenAt,
    this.latenessMinutes,
    this.wasPostponed = false,
  });

  factory IntakeHistoryEntry.fromJson(Map<String, dynamic> json) {
    return IntakeHistoryEntry(
      intakeId: json['intakeId'] as int,
      timing: (json['timing'] as String).toUpperCase(),
      scheduledTime: parseBackendDateTime(json['scheduledTime'])!,
      status: json['status'] as String,
      outcome: IntakeOutcome.fromApi(json['outcome'] as String?),
      takenAt: parseBackendDateTime(json['takenAt']),
      latenessMinutes: json['latenessMinutes'] as int?,
      wasPostponed: json['wasPostponed'] as bool? ?? false,
    );
  }
}
